inc/zoo/swar/SWAR.h

```diff
@@ -83,99 +88,153 @@ namespace detail {
     return 8*sizeof(T) - 1 - __builtin_clzll(v);
 }
 
-template<int Size, typename T = uint64_t> struct SWAR {
+template<typename T> constexpr typename std::make_unsigned<T>::type lsb_index(T v) noexcept {
+    return __builtin_ctzll(v) + 1;
+}
+
+template<int NBits, typename T = uint64_t> struct SWAR {
     SWAR() = default;
     constexpr explicit SWAR(T v): m_v(v) {}
+    constexpr explicit operator T() const noexcept { return m_v; }
 
     constexpr T value() const noexcept { return m_v; }
 
-    constexpr SWAR operator|(SWAR o) { return SWAR(m_v | o.m_v); }
-    constexpr SWAR operator&(SWAR o) { return SWAR(m_v & o.m_v); }
-    constexpr SWAR operator^(SWAR o) { return SWAR(m_v ^ o.m_v); }
+    constexpr SWAR operator|(SWAR o) const noexcept { return SWAR(m_v | o.m_v); }
+    constexpr SWAR operator&(SWAR o) const noexcept { return SWAR(m_v & o.m_v); }
+    constexpr SWAR operator^(SWAR o) const noexcept { return SWAR(m_v ^ o.m_v); }
 
-    constexpr bool operator==(SWAR o) { return m_v == o.m_v; }
-
-    constexpr T at(int position) {
-        constexpr auto filter = (T(1) << Size) - 1;
-        return filter & (m_v >> (Size * position));
+    constexpr T at(int position) const noexcept {
+        constexpr auto filter = (T(1) << NBits) - 1;
+        return filter & (m_v >> (NBits * position));
     }
 
-    constexpr SWAR clear(int position) {
-        constexpr auto filter = (T(1) << Size) - 1;
-        auto invertedMask = filter << (Size * position);
+    constexpr SWAR clear(int position) const noexcept {
+        constexpr auto filter = (T(1) << NBits) - 1;
+        auto invertedMask = filter << (NBits * position);
         auto mask = ~invertedMask;
         return SWAR(m_v & mask);
     }
 
-    constexpr int top() { return msb_index(m_v) / Size; }
-    constexpr int fastIndex() { return __builtin_ctzll(m_v) / Size; }
+    /// Notably, this is the SWAR lane index that contains the MSB.
+    constexpr int top() const noexcept { return msb_index(m_v) / NBits; }
+    /// Notably, this is the SWAR lane index that contains the lsb.
+    constexpr int lsbIndex() const noexcept { return __builtin_ctzll(m_v) / NBits; }
 
-    constexpr SWAR set(int index, int bit) {
-        return SWAR(m_v | (T(1) << (index * Size + bit)));
+    constexpr SWAR set(int index, int bit) const noexcept {
+        return SWAR(m_v | (T(1) << (index * NBits + bit)));
     }
 
-    constexpr operator bool() const { return m_v; }
-protected:
     T m_v;
 };
 
-template<int Size, typename T>
-struct BooleanSWAR: SWAR<Size, T> {
-    constexpr BooleanSWAR(T v): SWAR<Size, T>(v) {}
+/// Defining operator== on base SWAR types is entirely too error prone. Force a verbose invocation.
+template<int NBits, typename T = uint64_t> constexpr auto horizontalEquality(SWAR<NBits, T> left, SWAR<NBits, T> right) {
+  return left.m_v == right.m_v;
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto isolate(T pattern) {
+  return pattern & ((T(1)<<NBits)-1);
+}
```

Comment:
I've just discovered a serious error: imagine the element size is equal to swar size (this ought to be fully supported, BTW), then `1 << NBits` is either a no-op and then the code is consuming whatever garbage was there, or perhaps 0.  This case is not even *defined*.  We somehow have to make sure we **never** shift as many bits as the count of bits in the integer type.

Reply:
Yeah, I definitely noticed that edge case. Not exactly sure what to do about it.